File: src/main/java/com/saomc/saoui/social/party/DefaultParty.java
package com.saomc.saoui.social.party;

import com.saomc.saoui.api.social.party.IParty;
import net.minecraft.entity.player.EntityPlayer;

import java.util.ArrayList;
import java.util.List;

/**
 * Part of saoui
 * Default implementation of {@link IParty}.
 *
 * @author Bluexin
 */
public class DefaultParty implements IParty {
    private final List<EntityPlayer> members = new ArrayList<>(); // TODO: check if this can't cause a mem leak (using direct refs instead of WeakReferences)

    @Override
    public boolean addMember(EntityPlayer member) {
        return member != null && !this.members.contains(member) && this.members.add(member);
    }

    @Override
    public boolean removeMember(EntityPlayer member) {
        boolean flag = this.members.remove(member);
        if (this.size() <= 1) this.dissolve();
        return flag;
    }

    @Override
    public List<EntityPlayer> getMembers() {
        return this.members;
    }

    @Override
    public EntityPlayer getLeader() {
        return this.size() > 0 ? this.members.get(0) : null;
    }

    @Override
    public void dissolve() {
        this.members.clear();
    }

    @Override
    public int size() {
        return this.members.size();
    }

    @Override
    public boolean isParty() {
        return this.size() > 1;
    }

    @Override
    public boolean isInParty(EntityPlayer player) {
        return this.members.contains(player);
    }
}

File: src/main/java/com/saomc/saoui/social/party/DefaultParty.kt
package com.saomc.saoui.social.party

import com.saomc.saoui.api.social.party.IParty
import net.minecraft.entity.player.EntityPlayer

import java.util.ArrayList

/**
 * Part of saoui
 * Default implementation of [IParty].

 * @author Bluexin
 */
class DefaultParty : IParty {
    private val members = ArrayList<EntityPlayer>() // TODO: check if this can't cause a mem leak (using direct refs instead of WeakReferences)

    override fun addMember(member: EntityPlayer?): Boolean {
        return member != null && !this.members.contains(member) && this.members.add(member)
    }

    override fun removeMember(member: EntityPlayer): Boolean {
        val flag = this.members.remove(member)
        if (this.size() <= 1) this.dissolve()
        return flag
    }

    override fun getMembers(): List<EntityPlayer> {
        return this.members
    }

    override fun getLeader(): EntityPlayer? {
        return if (this.size() > 0) this.members[0] else null
    }

    override fun dissolve() {
        this.members.clear()
    }

    override fun size(): Int {
        return this.members.size
    }

    override fun isParty(): Boolean {
        return this.size() > 1
    }

    override fun isInParty(player: EntityPlayer): Boolean {
        return this.members.contains(player)
    }
}
